### ISSUE TITLE
Camera.getAvailableCameraTypesAsync() does not work on the web and native

### ISSUE THREAD

**Issue by kopax** [expo/expo#7501]
Please provide the following:
1. SDK Version: 36
2. Platforms(Android/iOS/web/all): all

I am working on a `react-native` application and I want to get the camera types (front and back) on the web and native.

In the [documentation](https://docs.expo.io/versions/latest/sdk/camera/#cameragetavailablecameratypesasync-string), it is written:

> `Camera.getAvailableCameraTypesAsync(): string[]`
Returns a list of camera types `['front', 'back']`. This is useful for desktop browsers which only have front-facing cameras.
>
> ```jsx
> import { Camera } from 'expo-camera';
> 
> const types = await Camera.getAvailableCameraTypesAsync();
> ```

This is my `Camera.js`:

```jsx
import React, { useState, useEffect } from 'react';
import { Platform, View, TouchableOpacity } from 'react-native';
import { FAB, Text } from 'react-native-paper';
import { Camera } from 'expo-camera';

export default function CoreCamera() {
  const [hasPermission, setHasPermission] = useState(null);
  const [type, setType] = useState(Camera.Constants.Type.back);
  const [types, setTypes] = useState(null);
  useEffect(() => {
    (async () => {
      const types = await Camera.getAvailableCameraTypesAsync();
      alert(JSON.stringify(types));
      setTypes(types);
      if (Platform.OS === 'web') {
        setHasPermission(true);
      } else {
        const { status } = await Camera.requestPermissionsAsync();
        setHasPermission(status === 'granted');
      }
    })();
  }, []);

  if (hasPermission === null) {
    return <View />;
  }
  if (hasPermission === false) {
    return <Text>No access to camera</Text>;
  }

  return (
    <View style={{ flex: 1 }}>
      <Camera style={{ flex: 1 }} type={type}>
        <View
          style={{
            flex: 1,
            backgroundColor: 'transparent',
            flexDirection: 'row',
          }}
        >
          <TouchableOpacity
            style={{
              flex: 0.1,
              alignSelf: 'flex-end',
              alignItems: 'center',
            }}
            onPress={() => {
              setType(
                type === Camera.Constants.Type.back
                  ? Camera.Constants.Type.front
                  : Camera.Constants.Type.back,
              );
            }}
          >
            <FAB
              style={{ marginBottom: 20 }}
              icon="camera-switch"
            />
          </TouchableOpacity>
        </View>
      </Camera>
    </View>
  );
}
```

On Android, it gives me the following error:

```jsx
[Unhandled promise rejection: Error: The method or property expo-camera.getAvailableCameraTypesAsync is not available on android, are you sure you've linked all the native dependencies properly?]
```

On iOS, it gives me the following error:

```jsx
[Unhandled promise rejection: Error: The method or property expo-camera.getAvailableCameraTypesAsync is not available on ios, are you sure you've linked all the native dependencies properly?]

```

On the web, it gives me the following error:

```jsx
bundle.js:73612 Uncaught (in promise) TypeError: Cannot read property 'getAvailableCameraTypesAsync' of undefined
    at getAvailableCameraTypesAsync$ (bundle.js:73612)
    at tryCatch (bundle.js:185959)
```

I have tried to replace:

```diff
-import { Camera } from 'expo-camera';
+import Camera from 'expo-camera/build/Camera';
```

Now the alert shows an empty array on the web.

How can I test the presence of the back camera on the web with expo?

expo-camera version 8.0.0

**edit**

I have tried to update `expo-camera@8.2.0` and it does the same.

**Comment by cruzach** [comment]
Looks like we need to update the documentation, that method is Web only (it's really only helpful on Web, since most smart phones have both front and back cameras)

As for your web error, could you share a repro? I tested it locally, on chrome, and didn't have any issues

**Comment by kopax** [comment]
Do you assume that all smartphone now have a front and a rear camera ? Because that's what I must do without the native support.

For the web, it does not work yes I confirm, I am willing to do a reproduction to move on this a little bit, do you think I can use `expo-camera` on snack.io or should I prepare a git repository ?

**Comment by cruzach** [comment]
Yeah, i think it's safe to assume that on native

You can create a repro on snack, which will automatically use the correct `expo-camera` version i think

**Comment by kopax** [comment]
It doesn't seems to work on snack using the exemple from the [doc](https://docs.expo.io/versions/v36.0.0/sdk/camera/#example-usage): https://snack.expo.io/@kopax/ashamed-apples

Should I prepare a reproduction on GitHub?

**Comment by kopax** [comment]
Good morning @cruzach, here is the reproduction (I mixed it with the reproduction for #7509 and Android not being able to take pictures.): 

```bash
git clone https://github.com/kopax/expo-reproduction-boilerplate.git
cd expo-reproduction-boilerplate
git checkout issues-7509-7501
yarn
npm run start
# Click on `w` to open your  web browser (I use Google Chrome)
```

1. Click on the camera tab in the bottom navigation
1. Grant camera permissions
1. Notice in the alert the following error:

```jsx
{"types":null,"error":"Cannot read property 'getAvailableCameraTypesAsync' of undefined"}
```

It was tested on the Samsung Galaxy S9, iPhone 6, and a browser.

**Comment by kopax** [comment]
> Anyway on Android it is essential to set ratio prop to get a picture with correct dimensions.

I would replace in the doc with this

> Anyway on Android it is **required** to set ratio prop to get a picture with correct dimensions.

There is a terminology widely used in the technical word.

With it, **it work on Android**, the example doesn't show this subtility, perhaps this can be improved, I leave it open for feedback.

**Comment by cruzach** [comment]
it looks like the problem is the permissions request, if you use `expo-permissions` it works - https://snack.expo.io/@charliecruzan/dc282b?platform=android&name=Basic%20Camera%20usage&sdkVersion=36.0.0&dependencies=expo-camera&sourceUrl=https%3A%2F%2Fdocs.expo.io%2Fstatic%2Fexamples%2Fv36.0.0%2Fcamera.js

**Comment by kopax** [comment]
> it looks like the problem is the permissions request, if you use `expo-permissions` it works - https://snack.expo.io/@charliecruzan/dc282b?platform=android&name=Basic%20Camera%20usage&sdkVersion=36.0.0&dependencies=expo-camera&sourceUrl=https%3A%2F%2Fdocs.expo.io%2Fstatic%2Fexamples%2Fv36.0.0%2Fcamera.js

@cruzach From my understanding this is a bug and should be corrected, either in the doc, or in the module. 

It can't just be close for others reading the same documentation as me.

**Comment by kopax** [comment]
@cruzach I have opened the link on the web and after installing `expo-permissions` it work

But this only concerns the permissions issue that I have solved on the web like follow:

```jsx
if (Platform.OS === 'web') {
   setHasPermission(true);
 }
```

Thanks for giving another option, but this should be corrected in the documentation as well or in the module. 

Also, this snack does not run `getAvailableCameraTypesAsync()`, which is the main discussion here. 

Closing an issue that can be resolved prior to it's resolution is an hostile action, I am saying this to you in consequence. I have prepared a reproduction for you, did you try to run it ? I must open a new issue. :/
